openmc/material.py

```diff
@@ -0,0 +1,1109 @@
+
+    next_id = 1
+    used_ids = set()
+
+    def __init__(self, material_id=None, name='', temperature=None):
+        # Initialize class attributes
+        self.id = material_id
+        self.name = name
+        self.temperature = temperature
+        self._density = None
+        self._density_units = 'sum'
+        self._depletable = False
+        self._paths = None
+        self._num_instances = None
+        self._volume = None
+        self._atoms = {}
+        self._isotropic = []
+
+        # A list of tuples (nuclide, percent, percent type)
+        self._nuclides = []
+
+        # The single instance of Macroscopic data present in this material
+        # (only one is allowed, hence this is different than _nuclides, etc)
+        self._macroscopic = None
+
+        # If specified, a list of table names
+        self._sab = []
+
+        # If true, the material will be initialized as distributed
+        self._convert_to_distrib_comps = False
+
+        # If specified, this file will be used instead of composition values
+        self._distrib_otf_file = None
+
+    def __repr__(self):
+        string = 'Material\n'
+        string += '{: <16}=\t{}\n'.format('\tID', self._id)
+        string += '{: <16}=\t{}\n'.format('\tName', self._name)
+        string += '{: <16}=\t{}\n'.format('\tTemperature', self._temperature)
+
+        string += '{: <16}=\t{}'.format('\tDensity', self._density)
+        string += ' [{}]\n'.format(self._density_units)
+
+        string += '{: <16}\n'.format('\tS(a,b) Tables')
+
+        for sab in self._sab:
+            string += '{: <16}=\t{}\n'.format('\tS(a,b)', sab)
+
+        string += '{: <16}\n'.format('\tNuclides')
+
+        for nuclide, percent, percent_type in self._nuclides:
+            string += '{: <16}'.format('\t{}'.format(nuclide))
+            string += '=\t{: <12} [{}]\n'.format(percent, percent_type)
+
+        if self._macroscopic is not None:
+            string += '{: <16}\n'.format('\tMacroscopic Data')
+            string += '{: <16}'.format('\t{}'.format(self._macroscopic))
+
+        return string
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def temperature(self):
+        return self._temperature
+
+    @property
+    def density(self):
+        return self._density
+
+    @property
+    def density_units(self):
+        return self._density_units
+
+    @property
+    def depletable(self):
+        return self._depletable
+
+    @property
+    def paths(self):
+        if self._paths is None:
+            raise ValueError('Material instance paths have not been determined. '
+                             'Call the Geometry.determine_paths() method.')
+        return self._paths
+
+    @property
+    def num_instances(self):
+        if self._num_instances is None:
+            raise ValueError(
+                'Number of material instances have not been determined. Call '
+                'the Geometry.determine_paths() method.')
```

Comment:
you briefly mention that these are determined by the `Geometry` methods in the Attributes class-doc section, but consider making this a stronger warning - that if you try to access these attrs without calling these methods then you'll get an exception - that wasn't what I was expecting when reading the docs.